src/app/TimedHandler.cpp

```diff
@@ -0,0 +1,138 @@
+        if (err != CHIP_NO_ERROR)
+        {
+            ChipLogError(DataManagement, "Failed to parse Timed Request action: handler %p exchange " ChipLogFormatExchange, this,
+                         ChipLogValueExchange(aExchangeContext));
+            StatusResponse::SendStatusResponse(Status::InvalidAction, aExchangeContext, /* aExpectResponse = */ false);
+        }
+        return err;
+    }
+
+    if (mState == State::kExpectingFollowingAction)
+    {
+        if (System::SystemClock().GetMonotonicTimestamp() > mTimeLimit)
+        {
+            // Time is up.  Spec says to send UNSUPPORTED_ACCESS.
+            ChipLogError(DataManagement, "Timeout expired: handler %p exchange " ChipLogFormatExchange, this,
+                         ChipLogValueExchange(aExchangeContext));
+            return StatusResponse::SendStatusResponse(Status::UnsupportedAccess, aExchangeContext, /* aExpectResponse = */ false);
+        }
+
+        if (aPayloadHeader.HasMessageType(MsgType::InvokeCommandRequest))
+        {
+            auto * imEngine = InteractionModelEngine::GetInstance();
+            aExchangeContext->SetDelegate(imEngine);
+            ChipLogDetail(DataManagement, "Handing timed invoke to IM engine: handler %p exchange " ChipLogFormatExchange, this,
+                          ChipLogValueExchange(aExchangeContext));
+            imEngine->OnTimedInvoke(this, aExchangeContext, aPayloadHeader, std::move(aPayload));
+            return CHIP_NO_ERROR;
+        }
+
+        // TODO: Add handling of MsgType::WriteRequest here.
+    }
+
+    // Not an expected message.  Send an error response.  The exchange will
+    // close when we return.
+    ChipLogError(DataManagement, "Unexpected unknown message in tiemd interaction: handler %p exchange " ChipLogFormatExchange,
+                 this, ChipLogValueExchange(aExchangeContext));
+
+    return StatusResponse::SendStatusResponse(Status::InvalidAction, aExchangeContext, /* aExpectResponse = */ false);
+}
+
+void TimedHandler::OnExchangeClosing(Messaging::ExchangeContext *)
+{
+    InteractionModelEngine::GetInstance()->OnTimedInteractionFailed(this);
+}
+
+CHIP_ERROR TimedHandler::HandleTimedRequestAction(Messaging::ExchangeContext * aExchangeContext,
+                                                  const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    using namespace Protocols::InteractionModel;
+
+    constexpr TLV::Tag timeoutTag = TLV::ContextTag(0);
+
+    // Parse the action, store things, etc.
+    System::PacketBufferTLVReader reader;
+    reader.Init(std::move(aPayload));
+
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::AnonymousTag));
+    // Enter the top-level TimedRequestMessage struct.
+    TLV::TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+
+    // Our timeout should be the first tag.
+    ReturnErrorOnFailure(reader.Next(timeoutTag));
```

Comment:
I think you can use https://github.com/project-chip/connectedhomeip/blob/master/src/app/MessageDef/TimedRequestMessage.cpp?

Reply:
I'll think about it, but it's missing some of the checks that I think we should be doing here, while doing some other checks multiple times for no particularly good reason.  And this parsing is simple enough it didn't seem worth pulling in all that machinery.  At the very least I'd want the machinery fixed to check the right things first.....